io/pci/pci_hotplug.py

```diff
@@ -128,16 +129,71 @@ def is_recovered():
             `pre` value, and returns True if equals pre.
             False otherwise.
             """
-            post = len(pci.get_disks_in_pci_address(pci_addr))
-            post += len(pci.get_nics_in_pci_address(pci_addr))
-            self.log.debug("Pre: %d,  Post: %d",
-                           self.end_devices[pci_addr], post)
-            if post == self.end_devices[pci_addr]:
+            def is_path_online():
+                path_stat = []
+                path_stat = list(multipath.get_path_status(curr_path))
+                self.log.info("path=%s status=%s" % (curr_path, path_stat))
+                if path_stat[0] != 'active' or path_stat[2] != 'ready':
+                    return False
+                return True
+
+            curr_path = ''
+            err_disks = []
+            if pci.get_pci_class_name(pci_addr) == 'fc_host':
+                disks = pci.get_disks_in_pci_address(pci_addr)
+                err_disks = []
+                for disk in disks:
+                    curr_path = disk.split("/")[-1]
+                    self.log.info("curr_path=%s" % curr_path)
+                    if not wait.wait_for(is_path_online, timeout=10):
+                        self.log.info("%s failed to recover after add" % disk)
+                        err_disks.append(disk)
+                    else:
+                        self.log.info("%s Recovered successfully" % disk)
+
+            if err_disks:
+                self.log.info("few paths failed to recover : %s" % err_disks)
+                return False
+            else:
+                self.log.info("All Paths recovered successfully")
                 return True
-            return False
 
-        if not wait.wait_for(is_added, timeout=10):
-            return False
-        # Waiting for 10s per end device, for recovery.
-        return wait.wait_for(is_recovered,
-                             self.end_devices[pci_addr] * 10)
+        def net_recovery_check():
+            """
+            calls different recovery functions based on adapter type
+            """
+            local = LocalHost()
+            iface = []
+            iface = pci.get_interfaces_in_pci_address(pci_addr, 'net')
```

Comment:
we can remove list initialzation as it already returns list or we can convert it to list using split()